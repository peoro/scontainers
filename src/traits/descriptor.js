
const assert = require('assert');
const straits = require('straits');
const utilTraits = require('./utils.js');

use traits * from straits.utils;

const toStr = straits.common.toString.*asMethod();

// metadata traits for Scontainers
const descriptorTraits = straits.utils.TraitSet.fromKeys({
	InnerCollection: Object,	// the container this container wraps
	innerCollectionKey: '',	// the key of the inner scontainers
	argKeys: [''],	// the names of the data-fields of this container

	mappingOnly: false,	// the decorator is just a map-like function
	standardIteration: false,	// iterating every element in order from `0` to `len()-1`
	transformStream: false,	// the decorator can transform the inner collection as a stream
});

use traits * from utilTraits;
use traits * from descriptorTraits;

Object.prototype.*describeScontainer = function( props ) {
	assert( props.argKeys, `${this}'s descriptor is missing argKeys` );
	assert( (!!props.innerCollectionKey)===(!!props.InnerCollection), `${this}'s descriptor needs to either have both innerCollectionKey(=${props.innerCollectionKey::toStr({verbose:0})}) and InnerCollection(=${props.InnerCollection::toStr({verbose:0})}) or none.` );

	// setting properties
	descriptorTraits.*implTraits( this, props );

	// setting default property values
	descriptorTraits.*addTraits( this, {
		mappingOnly: false,
		standardIteration: ( this.*mappingOnly || ! this.*InnerCollection ),
		transformStream: this.*mappingOnly,
	});
};

module.exports = descriptorTraits;
