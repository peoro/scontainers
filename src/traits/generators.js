
const straits = require('js-protocols');

const generatorTraits = straits.utils.TraitSet.fromKeys({
	// core protocol generators: static methods, explicitely implemented by a collection
	// these are used to dynamically generate core and derived protocols for higher performance

	// naturally-indexed collections (e.g. Array, Range)
	nthKVN( n ){},	// return the ${n}-th element - O(1)
	setNth( n, value ){},
	// needed to make naturally-indexed collections compatible with the associative collection API
	nToKey( n ){},	// return the key associated to the ${n}-th element (${get( nToKey(n) ) === nth(n)})
	keyToN( key ){},

	// associative collections (e.g. Map)
	// note that naturally-indexed collections are also associative, thanks to `nToKey` and `keyToN`
	getKVN( key ){},	// return the value of key ${key} - O(1)
	set( key, value ){},	// set ${value} as value for ${key} - O(1)
	hasKey( key ){},	// return true if ${this} has a key ${key}, false otherwise - O(1)

	//
	// delete( ? ){},	// deletes

	// unindexed collections (Set)
	has( item ){},	// return true or false depending on whether ${item} is in ${this} - O(1)

	// collections without a key, or with automatic key (e.g. Set, extendible naturally-indexed collections)
	add( value ){},	// adds ${value} to ${this} - O(1)

	// collections with known size
	len(){},	// return the number of elements in the collection - O(1)

	// clearable collection
	clear(){},	// removes every item from ${this} - O(n)

	// iterable collections
	loop(){},
});

module.exports = generatorTraits;
