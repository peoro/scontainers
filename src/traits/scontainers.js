
const straits = require('straits');

use traits * from straits.core;

// core traits
const coreTraits = straits.utils.TraitSet.fromKeys({
	// *** static methods
	from( collection ){},	// return a new instance of the collection constructed from `collection`

	// naturally-indexed collections (e.g. Array, Range)
	nth( n ){},	// return the ${n}-th element - O(1)
	nthKVN( n ){},	// NOTE: it doesn't check to see if `n` is there - O(1)
	setNth( n, value ){},	// sets the ${n}-th element to ${value} - O(1)
	truncate( n ){},	// shortens ${this} to contain the first ${n} elements only, discarding the others - O(1)
	// needed to make naturally-indexed collections compatible with the associative collection API:
	nToKey( n ){},	// return the key associated to the ${n}-th element (${get( nToKey(n) ) === nth(n)})
	keyToN( key ){},	// return the index associated to the element with key ${key} (${get(key) === nth(keyToN(key))})

	// associative collections (e.g. Map)
	// note that naturally-indexed collections are also associative, thanks to `nToKey` and `keyToN`
	get( key ){},	// return the value of key ${key} - O(1)
	getKVN( key ){},	// NOTE: it doesn't check to see if ${key} is there - O(1)
	set( key, value ){},	// set ${value} as value for ${key} - O(1)
	hasKey( key ){},	// return true if ${this} has a key ${key}, false otherwise - O(1)
	deleteKey( key ){},	// remove the item with key ${key} from ${this}

	// unindexed collections (e.g. Set, a list)
	has( item ){},	// return true or false depending on whether ${item} is in ${this} - O(1)
	delete( item ){},	// removes ${item} from ${this} - O(1)

	// collections without a key, or with automatic key (e.g. Set, naturally-indexed collections)
	add( value ){},	// adds ${value} to ${this} - O(1)

	// collections with known size
	len(){},	// return the number of elements in the collection - O(1)

	// reversible collections
	reverse(){},	// reverse iteration order - O(1)

	reserve( n ){},	// makes ${this} ready to host a total of ${n} items - O(1)

	clear(){},	// remove every item from ${this} - O(1)

	// iterable collections
	//iterator(){}, // not redefining it: reusing `[Symbol.Iterator]`
	kvIterator(){}, // identical to `[Symbol.Iterator]`, except the returned value is always in the form [key, value]
	kvReorderedIterator(){},
	// kvAsyncIterator(){}, // TODO: hmmmm...

	inplace(){},	// like ${consume}, but it must modify ${this} in place, otherwise it fails
});

// automatically implemented traits
const derivedTraits = straits.utils.TraitSet.fromKeys({
	// madifying functions
	// TODO: shouldn't these be core? they need to make assumptions on how to grow e.g. Array
	push( value ){},	// add ${value} to the end of ${this} - O(1)
	unshift( value ){},	// add ${value} to the beginning of ${this} - O(1)
	pop(){},	// remove the element at the end of ${this} - O(1)
	shift(){},	// remove the element at the beginning of ${this} - O(1)
	insert( n, value ){},	// insert ${value} after the ${n}-th element of ${this}
	remove( n ){},	// remove the ${n}-th element of ${this}, shifting all the following ones

	// iterating functions
	forEach( fn ){},	// call ${fn(value, key)} for every item in ${this} - O(n)
	whileEach( fn ){},	// - O(n)
	untilEach( fn ){},	// - O(n)

	// length information functions
	count(){},	// return the number of elements in the collection - any cost, often O(n)
	isEmpty(){},	// return true if the collection is empty, false otherwise - O(1)

	// special access functions - they all return KVNs
	only(){},	// return the only item in ${this} - O(1)
	first(){},	// return the first item in ${this} - O(1)
	last(){},	// return the last item in ${this} - O(1)
	random(){},	// return a random item in ${this} - O(n) // TODO: or O(1) ?
	find( fn ){},	// returning a KV
	findLast( fn ){},	// returning a KV

	swapNs( n1, n2 ){},	// swap two elements (by n)
	swapKeys( k1, k2 ){},	// swap two elements (by key)
	reduce( fn, initialValue ){},
	reduceFirst( fn ){},

	// reductions
	sum(){},
	avg(){},
	min(){},
	max(){},
	join( sep='' ){},	// returns a string
	every( fn ){},
	some( fn ){},

	sort( TargetArrayType=Array ){},
	shuffle(){},
	permute(){},

	collect( TargetType ){},	// creates a new ${TargetType} made of the content of ${this}
	consume( TargetType ){},	// like ${collect}, but might modify (and even return) ${this}

	// TODO: the following still need to be implemented
	remap( fn ){},	// [k:v]::remap( fn(v,k)->(k':v') ) -> [k':v']
	kvMap( fn ){},	// [k:v]::kvMap( fn(v,k)->v' ) -> [(k,v):v'] // .*mapKey( (v,k)=>{k,v} ).*map( (v,{k})=>fn(v,k) )
	unmap( fn ){},	// [(k,v)]::unmap() -> [k:v]
	unmapKeys( fn ){},	// .*keys().*unmap()
	allProperties(){},	// not iterable - manages all properties, owned or not, enumerable or not

	collectInto( target ){},
	repeat( n ){},
	// loop(){}, // repeat( Infinity )
});

const decoratorTraits = straits.utils.TraitSet.fromKeys({
	auto(){},	// after this, any operation will be available regardless of supported traits and complexity - if some are not supported, we automatically `collect` into a working type.
	keys(){},	// mostly needed because of `iterator`: iterates on a single value, rather than a KV
	values(){},	// mostly needed because of `iterator`: iterates on a single value, rather than a KV
	entries(){},	// mostly needed because of `iterator`: iterates on a KV
	properties(){},	// iterable: it's the enumerable properties
	ownProperties(){},	// the own properties of `this`
	enumerate(){},	// map( value => {count, value} )
	filter( fn ){},
	uniq(){},	// removes consecutive duplicates
	slice( begin, end ){},
	chunk( n ){},
	map( fn ){},
	mapKey(){},
	cache( CacheType ){},	// cache values resulting from `get()`, `nth()` etc. NOTE: it doesn't cache iteration!
	iter(){},	// removes other properties, leaves kvIterator only
	reordered(){},	// removes other properties, leaves kvReorderedIterator only
	cow( CopyType ){},	// copy on write: collects() as soon as you try to modify it
	groupBy( fn ){},
	flatten(){},
	flattenDeep(){},
	concat( ...collections ){},
	skipWhile( fn ){},
	takeWhile( fn ){},
	skip( n ){},
	take( n ){},
	assign( ...collections ){},
	defaults( ...collections ){},

	groupWhile(){},	// like `.*uniq`
});


derivedTraits.*assign({
	iterator: Symbol.iterator,
	toString: straits.common.toString,
});

const traits = ({}).*assign( derivedTraits, coreTraits, decoratorTraits );
traits.coreTraits = coreTraits;
traits.derivedTraits = derivedTraits;
traits.decoratorTraits = decoratorTraits;

module.exports = traits;

if( require.main === module ) {
	console.log( Object.keys(traits).join(`, `) );
}
