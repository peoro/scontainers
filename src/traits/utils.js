
const assert = require('assert');
const straits = require('straits');
const {id} = require('../utils_light.js');

use traits * from straits.core;
use traits * from straits.utils;

const utilTraits = straits.utils.TraitSet.fromKeys({
	extractKeys( keys ){},
	defaultGet( key, defaultConstructor ){},

	addTraits( target, implementationObj ){},
	addTraitFactories( target, factoryObj ){},
	implScontainer( implementationObj ){},
	wrapScontainer( decoratorFactoryFactory ){}, // return a function factory for the decorator `decoratorFactory` built on `this`

	describeScontainer(){},
	implCoreGenerators(){},
	implCoreTraits( ){},
	deriveTraits( ){},
});

utilTraits.*borrowTraits( straits.utils, [id`asFreeFunction`, id`asMethod`, id`impl`, id`implTraits`] );
utilTraits.*borrowTraits( straits.common, [id`toString`] );


use traits * from utilTraits;


// remove all the `keys` from `this`, and return an object made of all the removed keys and their values.
Object.prototype.*extractKeys = function( keys ) {
	const result = {};
	keys.forEach( key=>{
		result[key] = this[key];
		delete this[key];
	});
	return result;
};
Map.prototype.*defaultGet = function( key, defaultConstructor ) {
	if( this.has(key) ) {
		return this.get( key );
	} else {
		const value = defaultConstructor( key );
		this.set( key, value );
		return value;
	}
};

// like `implTraits`, but without overriding already implemented traits
Object.prototype.*addTraits = function( target, implementationObj ) {
	for( let name in implementationObj ) {
		const sym = this[name];
		assert( typeof sym === 'symbol', `No trait \`${name}\`` );

		if( target.hasOwnProperty(sym) ) {
			// trait already implemented; skipping
			continue;
		}

		sym.*impl( target, implementationObj[name] );
	};
}

class Factory {
	constructor( factory, target ) {
		this.target = target;

		if( ! factory ) {
			this.factory = function(){ throw new Error(`No factory`); };
			this.assignImmediately = false;
			this.canProduce = ()=>false;
		}
		else if( typeof factory === 'function' ) {
			this.factory = factory;
			this.assignImmediately = false;
			this.canProduce = function(){ return target::this.factory() !== undefined; };
		}
		else {
			this.factory = factory.factory;
			this.assignImmediately = factory.assignImmediately || false;
			this.canProduce = factory.canProduce || (()=>true);
		}
	}

	produce( ) {
		const {target} = this;
		const fn = target::this.factory();
		assert( fn !== undefined, `A factory didn't produce anything` );
		fn.factory = function(){ return fn; };
		return fn;
	}
}

// `factoryObj` has symbol names as traits and factories as
Object.prototype.*addTraitFactories = function( target, factoryObj ) {
	for( let name in factoryObj ) {
		const factory = new Factory( factoryObj[name], target );
		if( ! factory.canProduce() ) {
			// factory can't produce anything
			continue;
		}

		const sym = this[name];
		assert( typeof sym === 'symbol', `No trait \`${name}\`` );

		if( target.hasOwnProperty(sym) ) {
			// trait already implemented; skipping
			continue;
		}

		if( factory.assignImmediately ) {
			const fn = factory.produce();
			sym.*impl( target, fn );
		}
		else {
			const lazyFn = function() {
				const fn = factory.produce();
				sym.*impl( target, fn );
				return this[sym]( ...arguments );
			};
			lazyFn.factory = function() {
				const fn = factory.produce();
				sym.*impl( target, fn );
				return fn;
			};

			sym.*impl( target, lazyFn );
		}
	}
};

module.exports = utilTraits;

if( require.main === module ) {
	const traits = new straits.utils.TraitSet('a', 'b', 'c', 'x', 'y', 'z');
	use traits * from traits;

	const logs = [];
	function log( arg ) {
		logs.push( arg );
	}
	// checks and clears logs
	function checkLogs( ...expectedLogs ) {
		assert.deepStrictEqual( logs, expectedLogs );
		logs.length = 0;
	}

	const obj = {};
	const factoryObj = {
		a() {
			log( `a factory` );
			return function() {
				log( `a` );
			}
		},
		b() {
			log( `b factory` );
		},
		c: null,

		x: {
			assignImmediately: true,
			factory() {
				log( `x factory` );
				return function() {
					log( `x` );
				}
			},
		},
		y: {
			factory() {
				log( `y factory` );
				return function() {
					log( `y` );
				}
			},
		},
		z: {
			factory() {
				log( `z factory` );
				return function() {
					log( `z` );
				}
			},
			canProduce() {
				log( `z can produce` );
				return false;
			}
		},
	};

	// nothing was logged until now
	checkLogs();

	// so far the non-object factories were just tested to see whether they're valid
	traits.*addTraitFactories( obj, factoryObj );
	checkLogs(
		`a factory`,
		`b factory`,
		`x factory`,
		`z can produce`,
	);

	obj.*a();
	checkLogs( `a factory`, `a` );
	obj.*a();
	checkLogs( `a` );

	assert.throws( ()=>obj.*b() );
	assert.throws( ()=>obj.*c() );

	obj.*x();
	checkLogs( `x` );

	obj.*y();
	checkLogs( `y factory`, `y` );
	obj.*y();
	checkLogs( `y` );

	assert.throws( ()=>obj.*z() );
}
