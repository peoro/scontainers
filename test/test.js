
'use strict';

const scontainers = require('../src/index.js');
const {Range} = scontainers;

use traits * from scontainers;

Error.stackTraceLimit = Infinity;

function overrideFunctionName( name ) {
	Object.defineProperty( this, 'name', { get(){ return name; } });
}

function square( n ) { return n*n; }
square::overrideFunctionName( `n²` );

function multipleOf( k ) {
	function multipleOfK( n ) {
		return n % k === 0;
	};
	multipleOfK::overrideFunctionName( `n|${k}` );
	return multipleOfK;
}
function lessThan( k ) {
	function lessThanK( n ) {
		return n < k;
	};
	lessThanK::overrideFunctionName( `n<${k}` );
	return lessThanK;
}

function groupBy23( n ) {
	return (n%2===0) + (n%3===0);
}
groupBy23::overrideFunctionName( `n|2+n|3` );



{
	function print( str, coll ) {
		if( ! coll ) {
			coll = str;
			str = null;
		}
		if( str ) {
			console.log(str);
		}
		console.log( coll.toString() );
		console.log( coll.*toString() );
		console.log();
	}

	print(
		[1, 2, 3]
	);

	print(
		new Range( 3, 12 )
	);

	print(
		{ a:1, b:'hey' }.*properties()
	);

	print(
		[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
			.*slice(3, 7)
	);

	print(
		new Range(11)
			.*slice(3, 7)
	);

	print(
		new Range(21)
			.*chunk(5)
	);

	print(
		new Range(21)
			.*chunk(5)
			.*flatten()
			.*map( square )
	);

	print(
		new Range(2, 5).*map( square ).*map( square ).*map( square )
	);

	/*
	print(
		new Range(10)
			.*reordered()
			.*mapKey( (v,k)=>10-k )
			.*collect( Map )
	);
	process.exit( 0 );
	*/

	for( let kv of new Range(3).*map(n=>n+1).*iter().*map(n=>n*n) ) {
		console.log( kv, `...${kv[0]}: ${kv[1]}` );
	}
	console.log();

	const rit = new Range(3).*iter().*map(n=>n+1).*reordered().*map(n=>n*n).*kvReorderedIterator();
	rit.onNext( (kv)=>void console.log( kv, `...${kv.key}: ${kv.value}` ) )
		.proceed();
	console.log();

	{
		console.log( `GroupBy {` );
		console.group();

		const coll = new Range(1, 10)
			.*map( square )
			.*groupBy( groupBy23 );

		print( coll );

		{
			const rit = coll.*kvReorderedIterator();
			rit.onNext( (kv)=>{
				const groupKey = kv.key;
				const group = kv.value;
				console.log( `New group ${groupKey}` );

				const git = group
					// .*map( (v,k)=>v+k )
					.*kvReorderedIterator();

				git.onNext( (kv)=>{
					console.log( `  ${kv.key}:${kv.value} => group ${groupKey}` );
				})
				.proceed();
			})
			.proceed();
		}
		console.log();

		{
			const res = coll.*map( group=>{
				return group.*collect( Array );
			}).*collect( Map );

			console.log( res );
		}

		{
			const res = coll.*map( group=>{
				return group.*sum();
			}).*collect( Map );

			console.log( res );
		}
		{
			const res = coll.*map( group=>{
				return group.*sum();
			}).*sum()

			console.log( res );
		}
		console.groupEnd();
		console.log( `}` );
		console.log();
	}

	/*
	console.log(`yay`);
	process.exit( 0 );
	*/

	print( `0..21.chunk(5).flatten().map(x²).keys()`,
		new Range(21)
			.*chunk(5)
			.*flatten()
			.*map( square )
			.*keys()
	);

	print( `0..21.chunk(5).flatten().map(x²).values()`,
		new Range(21)
			.*chunk(5)
			.*flatten()
			.*map( square )
			.*values()
	);

	print( `0..21.chunk(5)[3]`,
		new Range(21)
			.*chunk(5)
			.*nth( 3 )
	);

	print(
		new Range(50)
			.*map( square )
			.*slice( 10, 40 )
			.*filter( multipleOf(2) )
			.*filter( multipleOf(3) )
			.*chunk( 3 )
	);

	print( `[...].flattenDeep().collect(Array)`,
		[ new Range(3), ['x', new Map([['x','x1'], ['y',['y2', 'y2b', 'y2c']], ['z','z3']]), 'z'], 'wow']
			.*flattenDeep()
			.*collect( Array )
	);

	const fib = new Range()
		.*map( function fibonacci(i){
			switch( i ) {
				case 0: return 0;
				case 1: return 1;
				default: return fib.*nth(i-2) + fib.*nth(i-1);
			}
		})
		.*cache(Array);

	print( `fib.take( 10 )`,
		fib.*take( 10 )
	);

	print( `fib.takeWhile( x<100 )`,
		fib.*takeWhile( lessThan(100) )
	);

	print( `fib.skipWhile( x<100 ).takeWhile( x<1000 )`,
		fib
			.*skipWhile( lessThan(100) )
			.*takeWhile( lessThan(1000) )
	);

	console.log();

	console.log( new Range(1e10+1).*avg(), new Range(1e10+1).*sum() );

	for( let r of new Range(5,10) ) {
		console.log( r );
	}
	for( let r of new Map([['x',1], [2,'y']]).*values() ) {
		console.log( r );
	}


	console.log( fib.*nth(70) );
}
