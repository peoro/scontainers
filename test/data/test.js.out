1,2,3
[1, 2, 3]

3…12
*[3, 4, 5, 6, 7, 8, 9, 10, 11]

{a:1, b:"hey"}::properties()
*{"a":1, "b":"hey"}

0,1,2,3,4,5,6,7,8,9,10[3:7]
*[3, 4, 5, 6]

3…7
*[3, 4, 5, 6]

0…21.chunk(5)
*[Map{0:0, 1:1, 2:2, 3:3, 4:4}, Map{5:5, 6:6, 7:7, 8:8, 9:9}, Map{10:10, 11:11, 12:12, 13:13, 14:14}, Map{15:15, 16:16, 17:17, 18:18, 19:19}, Map{20:20}]

0…21.chunk(5).flatten().map(n²)
*{0:0, 1:1, 2:4, 3:9, 4:16, 5:25, 6:36, 7:49, 8:64, 9:81, 10:100, 11:121, 12:144, 13:169, 14:196, 15:225, 16:256, 17:289, 18:324, 19:361, 20:400}

2…5.map(n²).map(n²).map(n²)
*[256, 6561, 65536]

[ 0, 1 ] '...0: 1'
[ 1, 4 ] '...1: 4'
[ 2, 9 ] '...2: 9'

KVN { key: 0, value: 1, n: 0 } '...0: 1'
KVN { key: 1, value: 4, n: 1 } '...1: 4'
KVN { key: 2, value: 9, n: 2 } '...2: 9'

GroupBy {
  1…10.map(n²).groupBy(n|2+n|3)
  *{2:*{6:36}, 1:*{2:4, 3:9, 4:16, 8:64, 9:81}, 0:*{1:1, 5:25, 7:49}}
  
  New group 0
    1:1 => group 0
  New group 1
    2:4 => group 1
    3:9 => group 1
    4:16 => group 1
    5:25 => group 0
  New group 2
    6:36 => group 2
    7:49 => group 0
    8:64 => group 1
    9:81 => group 1
  
  Map { 2 => [ 36 ], 1 => [ 4, 9, 16, 64, 81 ], 0 => [ 1, 25, 49 ] }
  Map { 2 => 36, 1 => 174, 0 => 75 }
  285
}

0..21.chunk(5).flatten().map(x²).keys()
0…21.chunk(5).flatten().map(n²).map(key).values()
*{0:0, 1:1, 2:2, 3:3, 4:4, 5:5, 6:6, 7:7, 8:8, 9:9, 10:10, 11:11, 12:12, 13:13, 14:14, 15:15, 16:16, 17:17, 18:18, 19:19, 20:20}

0..21.chunk(5).flatten().map(x²).values()
0…21.chunk(5).flatten().map(n²).values()
*{0:0, 1:1, 2:4, 3:9, 4:16, 5:25, 6:36, 7:49, 8:64, 9:81, 10:100, 11:121, 12:144, 13:169, 14:196, 15:225, 16:256, 17:289, 18:324, 19:361, 20:400}

0..21.chunk(5)[3]
15…20
*[15, 16, 17, 18, 19]

0…50.map(n²)[10:40].filter(n|2).filter(n|3).chunk(3)
*{0:Map{12:144, 18:324, 24:576}, 1:Map{30:900, 36:1296}}

[...].flattenDeep().collect(Array)
0,1,2,x,x1,y2,y2b,y2c,z3,z,wow
[0, 1, 2, "x", "x1", "y2", "y2b", "y2c", "z3", "z", "wow"]

fib.take( 10 )
0…Infinity.map(fibonacci).cache(Array)[0:10]
*[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

fib.takeWhile( x<100 )
0…Infinity.map(fibonacci).cache(Array).takeWhile(n<100)
*{0:0, 1:1, 2:1, 3:2, 4:3, 5:5, 6:8, 7:13, 8:21, 9:34, 10:55, 11:89}

fib.skipWhile( x<100 ).takeWhile( x<1000 )
0…Infinity.map(fibonacci).cache(Array).skipWhile(n<100).takeWhile(n<1000)
*{12:144, 13:233, 14:377, 15:610, 16:987}


5000000000 50000000005000000000
5
6
7
8
9
1
y
190392490709135
